src/USER-HSMA/hsma2d.cpp

```diff
@@ -0,0 +1,2056 @@
+
+		int ier = 0;
+		lfmm3d_t_c_g_(&eps, &ns, source, charge, &nt, target, pottarg, gradtarg, &ier);
+
+		for (int i = 0; i < S * Nw; i++)
+			for (int j = 0; j < S * Nw; j++)
+			{
+				Top[i][j] = pottarg[i * (S * Nw) + j];
+				TopZD[i][j] = gradtarg[3 * (i * (S * Nw) + j) + 2];
+				Down[i][j] = pottarg[(S * Nw) * (S * Nw) + i * (S * Nw) + j];
+				DownZD[i][j] = gradtarg[3 * ((S * Nw) * (S * Nw) + i * (S * Nw) + j) + 2];
+			}
+		
+		free(source); free(target); free(charge); free(pottarg); free(gradtarg);
+	}
+	else
+	{
+		double Paramet1[ImageNumber];
+		for (int i = 0; i < ImageNumber; i++)
+		{
+			Paramet1[i] = ImageCharge[i][3] * pow(Gamma, fabs(ImageCharge[i][4]) + 0.00);
+		}
+		
+		#pragma omp parallel
+		{
+			int id = omp_get_thread_num();
+			int size = omp_get_num_threads();
+			int min_atom=id*floor(S*Nw*S*Nw/size)+1, max_atom=(id+1)*floor(S*Nw*S*Nw/size);
+			if (id == size - 1)max_atom = S * Nw * S * Nw - 1;
+			if (id == 0)min_atom = 0;
+
+			int float_double;
+			if (tolerance > 0.000001) {
+				float_double = 1;
+				float IntegralTop_X[int(ceil( (max_atom - min_atom + 1)/16.0 ))*16], IntegralTop_Y[int(ceil((max_atom - min_atom + 1) / 16.0)) * 16], IntegralTop_Z[int(ceil((max_atom - min_atom + 1) / 16.0)) * 16];
+				for (int i = min_atom; i <= max_atom; i++)
+				{
+					int ix = floor(i / (S * Nw));
+					int iy = i - ix * (S * Nw);
+					IntegralTop_X[i - min_atom] = IntegralTop[ix][iy][0];
+					IntegralTop_Y[i - min_atom] = IntegralTop[ix][iy][1];
+					IntegralTop_Z[i - min_atom] = IntegralTop[ix][iy][2];
+				}
+
+				for (int i = max_atom - min_atom + 1; i<int(ceil((max_atom - min_atom + 1) / 16.0)) * 16;i++)
+				{
+					IntegralTop_X[i] = 0.00;
+					IntegralTop_Y[i] = 0.00;
+					IntegralTop_Z[i] = 0.00;
+				}
+
+				for (int i = min_atom; i <= max_atom; i = i + 16)
+				{
+					__m512 pottarg, fldtarg, pottarg2, fldtarg2, X0, Y0, X1, Y1, dx, dy, dz, dz1, delta, delta1, Para, midterm, midterm1;
+					float F1[16], F2[16], F3[16], F4[16];
+					pottarg = fldtarg = pottarg2 = fldtarg2 = _mm512_setzero_ps();
+					X0 = _mm512_load_ps(&IntegralTop_X[i - min_atom]);
+					Y0 = _mm512_load_ps(&IntegralTop_Y[i - min_atom]);
+
+					for (int j = 0; j < ImageNumber; j++)
+					{
+						Para = _mm512_set1_ps(Paramet1[j]);
+						X1 = _mm512_set1_ps(ImageCharge[j][0]);
+						Y1 = _mm512_set1_ps(ImageCharge[j][1]);
+						dz = _mm512_set1_ps(Lz / 2.0 - ImageCharge[j][2]);
+						dx = X0 - X1;
+						dy = Y0 - Y1;
+						delta = _mm512_invsqrt_ps(dx * dx + dy * dy + dz * dz);
+						midterm = Para * delta;
+						fldtarg -= dz * midterm * delta * delta;
+						pottarg += midterm;
+						dz1 = _mm512_set1_ps(-Lz / 2.0 - ImageCharge[j][2]);
+						delta1= _mm512_invsqrt_ps(dx * dx + dy * dy + dz1 * dz1);
+						midterm1 = Para * delta1;
+						pottarg2 += midterm1;
+						fldtarg2 -= dz1 * midterm1 * delta1 * delta1;
+					}
+
+					_mm512_store_ps(&F1[0], pottarg);
```

Comment:
Please do not use vector intrinsics directly. We try to keep LAMMPS portable to a large variety of platforms and compilers and these are highly non-portable. They can be hidden in macros, for example. See USER-INTEL package.

Reply:
Revised